### PR TITLE
backoffice: ensure items with active loans cannot change status

### LINKED ISSUE #182: Change item status should check for linked loans
If an item has `LOANABLE` status and it is changed to something else (for example `MISSING`) then the system should check if there is any active or pending loan on it and in this case:

* BLOCK the change, the active loan should be cancelled first.
**or**
* invalidate any active loan?

### PR DESCRIPTION
**Pull request by vlad-bm**
Ensure items with active loans cannot change status.
(Closes #182)

### CODE REVIEW

File: invenio_app_ils/records/api.py [hunk 1]
```diff
@@ -212,6 +213,19 @@ def create(cls, data, id_=None, **kwargs):
         }
         return super(Item, cls).create(data, id_=id_, **kwargs)
 
+    def ensure_item_can_be_updated(self):
+        """Raises an exception if the item's status cannot be updated."""
+        loan_search = current_circulation.loan_search
+        active_loans = loan_search\
+            .get_active_loans_by_item_pid(self["item_pid"]).execute().hits
+        if self["status"] == "CAN_CIRCULATE" and active_loans.total != 0:
```
Comment (zzacharo) [inline invenio_app_ils/records/api.py]: total > 0 ? I don't know if for some reason the total can be null...

**Review by zzacharo — DISMISSED** [review]

File: invenio_app_ils/circulation/search.py [hunk 2]
```diff
@@ -53,6 +53,14 @@ def get_loan_next_available_date(self, document_pid):
                              sort_by_field="end_date",
                              )
 
+    def get_active_loan_by_item_pid(self, item_pid):
+        """Return any active loans for the given item."""
+        return search_by_pid(item_pid=item_pid,
```
Comment (topless) [inline invenio_app_ils/circulation/search.py]: alignment
```
return search_by_pid(
    item_pid=item_pid,
    filter_states=current_app.config.get(
        "CIRCULATION_STATES_LOAN_ACTIVE", []
    ),
)                             
```

File: invenio_app_ils/records/api.py [hunk 3]
```diff
@@ -212,6 +213,19 @@ def create(cls, data, id_=None, **kwargs):
         }
         return super(Item, cls).create(data, id_=id_, **kwargs)
 
+    def ensure_item_can_be_updated(self):
+        """Raises an exception if the item's status cannot be updated."""
+        loan_search = current_circulation.loan_search
+        active_loan = loan_search\
+            .get_active_loan_by_item_pid(self["item_pid"]).execute().hits
```
Comment (topless) [inline invenio_app_ils/records/api.py]: Shouldn't this `self["item_pid"]` be  `self[Item.pid_field]`

File: tests/api/test_item_update.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import json
+
+from flask import url_for
+from invenio_accounts.models import User
+from invenio_accounts.testutils import login_user_via_session
+from invenio_circulation.proxies import current_circulation
+
+from invenio_app_ils.errors import ItemHasActiveLoansError
+
+NEW_LOAN = {
```
Comment (topless) [inline tests/api/test_item_update.py]: Is this used anywhere?

File: invenio_app_ils/errors.py [hunk 5]
```diff
@@ -97,6 +97,20 @@ def __init__(self, record_type, record_id, ref_type, ref_ids, **kwargs):
         )
 
 
+class ItemHasActiveLoansError(IlsException):
```
Comment (topless) [inline invenio_app_ils/errors.py]: Missing the test for this one in `test_errors.py`

File: tests/api/test_item_update.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        client,
+        email=User.query.get(users["admin"].id).email
+    )
+    item_pid = _get_item_on_loan()['item_pid']
+    patch_op = [{"op": "replace", "path": "/status", "value": "MISSING"}]
+    url = url_for('invenio_records_rest.pitmid_item', pid_value=item_pid)
+    res = client.patch(url,
+                       headers=json_patch_headers,
+                       data=json.dumps(patch_op))
+    assert res.status_code == ItemHasActiveLoanError.code
```
Comment (kpsherva) [inline tests/api/test_item_update.py:54]: check here also message string of the exception, because all the ILSException have the same 400 code

**Comment by zzacharo** [comment]
@vladbmihaescu your tests are failing because wrong order in the imports. Run locally the command `run_tests.sh` in order to fix the ordering and push again so we can merge it!

**Pull request merged**